src/Core/Worker.php

```diff
@@ -835,220 +820,188 @@ private static function activeWorkers()
 	}
 
 	/**
-	 * @brief Check if we should pass some slow processes
+	 * @brief Returns waiting jobs for the current process id
 	 *
-	 * When the active processes of the highest priority are using more than 2/3
-	 * of all processes, we let pass slower processes.
-	 *
-	 * @param string $highest_priority Returns the currently highest priority
-	 * @return bool We let pass a slower process than $highest_priority
+	 * @return array waiting workerqueue jobs
 	 * @throws \Exception
 	 */
-	private static function passingSlow(&$highest_priority)
+	private static function getWaitingJobForPID()
 	{
-		$highest_priority = 0;
-
 		$stamp = (float)microtime(true);
-		$r = DBA::p(
-			"SELECT `priority`
-				FROM `process`
-				INNER JOIN `workerqueue` ON `workerqueue`.`pid` = `process`.`pid` AND NOT `done`"
-		);
+		$r = DBA::select('workerqueue', [], ['pid' => getmypid(), 'done' => false]);
 		self::$db_duration += (microtime(true) - $stamp);
-
-		// No active processes at all? Fine
-		if (!DBA::isResult($r)) {
-			return false;
-		}
-		$priorities = [];
-		while ($line = DBA::fetch($r)) {
-			$priorities[] = $line["priority"];
+		if (DBA::isResult($r)) {
+			return DBA::toArray($r);
 		}
 		DBA::close($r);
 
-		// Should not happen
-		if (count($priorities) == 0) {
-			return false;
-		}
-		$highest_priority = min($priorities);
+		return false;
+	}
 
-		// The highest process is already the slowest one?
-		// Then we quit
-		if ($highest_priority == PRIORITY_NEGLIGIBLE) {
-			return false;
+	/**
+	 * @brief Returns the next jobs that should be executed
+	 *
+	 * @return array array with next jobs
+	 * @throws \Exception
+	 */
+	private static function nextProcess()
+	{
+		$priority = self::nextPriority();
+		if (empty($priority)) {
+			Logger::log('No tasks found', Logger::DEBUG);
+			return [];
 		}
-		$high = 0;
-		foreach ($priorities as $priority) {
-			if ($priority == $highest_priority) {
-				++$high;
-			}
+
+		if ($priority <= PRIORITY_MEDIUM) {
+			$limit = Config::get('system', 'worker_fetch_limit', 1);
+		} else {
+			$limit = 1;
 		}
-		Logger::log("Highest priority: ".$highest_priority." Total processes: ".count($priorities)." Count high priority processes: ".$high, Logger::DEBUG);
-		$passing_slow = (($high/count($priorities)) > (2/3));
 
-		if ($passing_slow) {
-			Logger::log("Passing slower processes than priority ".$highest_priority, Logger::DEBUG);
+		$ids = [];
+		$stamp = (float)microtime(true);
+		$condition = ["`priority` = ? AND `pid` = 0 AND NOT `done` AND `next_try` < ?", $priority, DateTimeFormat::utcNow()];
+		$tasks = DBA::select('workerqueue', ['id'], $condition, ['limit' => $limit, 'order' => ['created']]);
+		self::$db_duration += (microtime(true) - $stamp);
+		while ($task = DBA::fetch($tasks)) {
+			$ids[] = $task['id'];
 		}
-		return $passing_slow;
+		DBA::close($tasks);
+
+		Logger::log('Found task(s) ' . implode(', ', $ids) . ' with priority ' .$priority, Logger::DEBUG);
```

Comment:
Please use the new `Logger::error()`, `Logger::warning()`, `Logger::notice()`, `Logger::info()` and `Logger::debug()` methods at least for new logger calls.

Additionally, instead of interpolating variables in the log message, please add them to the `$context` array parameter of those methods instead.

Reply:
I must confess, that I totally dislike the concept of adding those variables at the end. It's like going to some restaurant and saying: "Hello, I would like with and some to drink. (dish: Steak, side: french fries, drink: soda)". What is the benefit? These messages should be human readable.

Reply:
It's more like "Hello, I would like something to eat and drink." and than the context what. So you can write text which is interoperable (f.e. "Worker Start") which is easily searchable and grep the context what needed. And you can use the context for specific logging. It should avoid mixing context up to heterogeneous outputs which are more human readable, but hard to grep per pattern.

And if we do have big instances, where f.e. is running a "loggrepper" or something like a log-database (f.e. splunk), the context would be used for indexing and search. 

But there's the way to combine both:
"Found task(s): {tasks}", ['tasks' => $tasks]" which will replace the placeholder with the content of the context

A good example is the "worker_id", which should be added at every worker output and has therefore the same type of output ("[worker_id => 123456]").

Reply:
Funnily after the changes in the logging I now do have much more problems finding stuff in the logfile at all - which is surely based upon the fact that I'm much more a developer than an administrator and I'm struggling with complex search patterns in "grep" (In fact I never heard of things like "splunk" or "loggrepper" at all). Currently I'm partially lost and with every of these changes, the log gets more unreadable for me. I guess I will need some tutorial how to use this changed log format in an efficient way. It would be good if someone could write such a tutorial and would place it in our documentation.

I will have to have a look at it. But this will surely last several days. I'm currently preparing the moving of one of my services to a new server. This has got the highest priority for me now.

Reply:
I'm sorry :-/ .. I'll try to add a little howto for log-searches ..

And I'll create a PR for switching from "Monolog" to "Syslog", which will dump the log in a more human readable way. Or at least two different context-possibilities ... I've to think about it ..

Reply:
BTW: This wasn't meant as an accusation (just to make this clear!). I just think that we should provide some help in using the tools as well, since there are most likely not so much professional administrators under the Friendica admins.

Whatever we will use as a log format in the future: I really do like the ones that aren't using too much space, since I often watch specific log entries with ```tail -f friendica.log | grep ...```. And there the multi line log entries don't look that nice.

Reply:
I personally like the new style even if it is slightly more verbose for the reason @nupplaphil: It is harder for me to grep the log lines I want if they have conditional info inside their message. On the other hand, I can add as much context data to the logger calls I'm adding without making the message unreadable.

Reply:
@annando will you add these new logger call to the PR? Just making sure no squirrel passed the way of the changes. ;-)

Reply:
I hadn't got time yet to have a deep look at it. I guess I will need some time to rewrite these logger calls to adopt to this new way, but to keep them readable and short.